File: src/fish/old_java/Sardin.java
/**
 * 
 */
package org.ahahah.fish.client;

import java.util.ArrayList;

/**
 * @author julien Gaffuri
 *
 */
public class Sardin {
	final static double D_OBS=50;
	final static double A_OBS=200*Math.PI/180;
	final static double D_COL=10;
	final static double V_TARGET=0.3;
	final static double V_MAX=0.6;

	Sea sea;
	double x=0,y=0; //position, in pix
	double vx=0,vy=0; //speed, in pix per ms
	double v=0, va=0; //speed angle (in -Pi,Pi) and norm
	double ax=0, ay=0; //acceleration, in pix per s2

	public Sardin(Sea sea) {
		this(sea, sea.w*Math.random(), sea.h*Math.random(), 0, 0);
		va=Math.random()*2*Math.PI-Math.PI;
		v=Math.random()*V_MAX;
		vx=v*Math.cos(va);
		vy=v*Math.sin(va);
	}

	public Sardin(Sea sea, double x, double y, double vx, double vy){
		this.sea=sea;

		if(x<0) this.x=0;
		else if(x>sea.w) this.x=sea.w;
		else this.x=x;

		if(y<0) this.y=0;
		else if(y>sea.h) this.y=sea.h;
		else this.y=y;

		this.vx=vx;
		this.vy=vy;
		v=Math.hypot(vx,vy);
		va=Math.atan2(vy,vx);

		sea.grid.add(this, this.x, this.y);
	}





	ArrayList<Sardin> obs = new ArrayList<Sardin>();
	ArrayList<Sardin> col = new ArrayList<Sardin>();
	public void observe() {
		ArrayList<Sardin> ss = sea.grid.get(x-D_OBS, y-D_OBS, x+D_OBS, y+D_OBS);

		obs.clear();
		col.clear();
		ArrayList<Sardin> col = new ArrayList<Sardin>();
		for(Sardin s:ss){
			if(s==this) continue;
			if(d(s)<=D_COL) col.add(s);
			if(d(s)<=D_OBS) {
				//check angle
				double da=Math.atan2(s.y-y,s.x-x)-va;
				if(da>Math.PI) da-=2*Math.PI;
				else if(da<=-Math.PI) da+=2*Math.PI;
				da=Math.abs(da);
				if(da>A_OBS*0.5) continue;
				obs.add(s);
			}
		}

		ax=0; ay=0;

		//repulsion
		for(Sardin s:col){
			double d=d(s);
			double a=1.0*(1/(d*d)-1/(D_COL*D_COL));
			ax+=a*(x-s.x)/d;
			ay+=a*(y-s.y)/d;
		}

		//toward v target
		double dv=(V_TARGET-v)*0.01;
		ax+=dv*vx/v;
		ay+=dv*vy/v;

		//toward the observed mean position
		/*if(obs.size()>1){
			double xn=0,yn=0;
			for(Sardin s:obs){
				xn+=s.x;
				yn+=s.y;
			}
			xn=xn/obs.size();
			yn=yn/obs.size();
			double d=Math.hypot(xn-x,yn-y);
			ax+=-0.1*(x-xn)/d;
			ay+=-0.1*(y-yn)/d;
		}*/

		//toward the observed speed
		double t=0.9;
		if(obs.size()>1){
			double dvx=0,dvy=0;
			for(Sardin s:obs){
				dvx+=s.vx;
				dvy+=s.vy;
			}
			dvx=(t-1)*vx+(1-t)*dvx/obs.size();
			dvy=(t-1)*vy+(1-t)*dvy/obs.size();

			double a=0.1;
			ax+=a*dvx;
			ay+=a*dvy;
		}

		//avoid shark
		if(sea.shark!=null){
			double[] s=sea.shark;
			double d=Math.hypot((s[0]-x), (s[1]-y));
			if(d<=D_OBS){
				double a=5.0*(1/(d*d)-1/(D_OBS*D_OBS));
				ax+=a*(x-s[0])/d;
				ay+=a*(y-s[1])/d;
			}
		}

	}

	public void move() {
		sea.grid.remove(this, this.x, this.y);

		//compute new speed
		vx+=ax*sea.timeStepMs +(1-2*Math.random())*0.02;
		vy+=ay*sea.timeStepMs +(1-2*Math.random())*0.02;
		v=Math.hypot(vx,vy);
		va=Math.atan2(vy, vx);
		if(v>V_MAX){
			v=V_MAX;
			vx=V_MAX*Math.cos(va);
			vy=V_MAX*Math.sin(va);
		}

		//compute new position
		x+=vx*sea.timeStepMs;
		y+=vy*sea.timeStepMs;

		//limit
		if(x<0) x=sea.w;
		if(y<0) y=sea.h;
		if(x>sea.w) x=0;
		if(y>sea.h) y=0;

		sea.grid.add(this, this.x, this.y);
	}

	public double d(Sardin s){
		return Math.hypot((s.x-x), (s.y-y));
	}
}

File: src/fish/Sardin.js
//@ts-check
import { Sea } from "./Sea"
import { GeoViewer } from "../base/GeoViewer"
import { AgentPoint } from "../base/AgentPoint";
import { SpatialIndex } from '../base/SpatialIndex';

/** */
export class Sardin extends AgentPoint {

    /**
     * @param {Sea} sea 
     * @param {number} x 
     * @param {number} y 
     * @param {number} sx 
     * @param {number} sy 
     */
    constructor(sea, x = sea.w * Math.random(), y = sea.h * Math.random(), sx = undefined, sy = undefined) {

        //
        super(x, y, sx, sy)

        /** @type {Sea} */
        this.sea = sea;

        //position
        /** @type {number} */
        this.x = x < 0 ? 0 : x > sea.w ? sea.w : x
        /** @type {number} */
        this.y = y < 0 ? 0 : y > sea.h ? sea.h : y

        //speed
        if (!sx && !sy) {
            this.setRandomSpeed(0, this.sea.V_MAX)

            //acceleration
            /** @type {number} */
            this.ax = 0
            /** @type {number} */
            this.ay = 0

            //list of sardins in vision field
            /** @type {Array.<Sardin>} */
            this.obs = []
            //list of sardins in collision field
            /** @type {Array.<Sardin>} */
            this.col = []
        }
    }


    /**
     * @param {SpatialIndex.<Sardin>} sindex 
     */
    observe(sindex) {

        /** @type {Sea} */
        const s = this.sea
        /** @type {number} */
        const dO = s.D_OBS

        //initialise lists
        /** @type {Array.<Sardin>} */
        this.obs = [];
        /** @type {Array.<Sardin>} */
        this.col = [];

        //get sardins around using spatial index
        /** @type {Array.<Sardin>} */
        const ss = sindex.get(this.x - dO, this.y - dO, this.x + dO, this.y + dO);

        //get sardins in observation and collision fields
        for (let f of ss) {
            if (f == this) continue;
            if (this.d(f) <= s.D_COL)
                this.col.push(f);
            if (this.d(f) <= dO) {
                //check angle
                let da = Math.atan2(f.y - this.y, f.x - this.x) - this.sa;
                if (da > Math.PI) da -= 2 * Math.PI;
                else if (da <= -Math.PI) da += 2 * Math.PI;
                da = Math.abs(da);
                if (da > s.A_OBS * 0.5) continue;
                this.obs.push(f);
            }
        }

        //initialise acceleration
        this.ax = 0; this.ay = 0;

        //collision: repulsion
        for (let f of this.col) {
            const d = this.d(f);
            const a = 1.0 * (1 / (d * d) - 1 / (s.D_COL * s.D_COL));
            this.ax += a * (this.x - f.x) / d;
            this.ay += a * (this.y - f.y) / d;
        }

        // toward target speed
        const dv = (s.V_TARGET - this.s) * 0.01;
        this.ax += dv * this.sx / this.s;
        this.ay += dv * this.sy / this.s;

        //toward the observed mean position
        /*if(obs.size()>1){
            double xn=0,yn=0;
            for(Sardin s:obs){
                xn+=s.x;
                yn+=s.y;
            }
            xn=xn/obs.size();
            yn=yn/obs.size();
            double d=Math.hypot(xn-x,yn-y);
            ax+=-0.1*(x-xn)/d;
            ay+=-0.1*(y-yn)/d;
        }*/

        //toward the observed speed
        const t = 0.9;
        if (this.obs.length > 1) {
            let dsx = 0, dsy = 0;
            for (let s of this.obs) {
                dsx += s.sx;
                dsy += s.sy;
            }
            dsx = (t - 1) * this.sx + (1 - t) * dsx / this.obs.length;
            dsy = (t - 1) * this.sy + (1 - t) * dsy / this.obs.length;

            const a = 0.1;
            this.ax += a * dsx;
            this.ay += a * dsy;
        }

        //avoid shark
        const sh = s.shark;
        if (sh != null) {
            const d = Math.hypot((sh.x - this.x), (sh.y - this.y));
            if (d <= s.D_OBS) {
                const a = 5.0 * (1 / (d * d) - 1 / (dO * dO));
                this.ax += a * (this.x - sh.x) / d;
                this.ay += a * (this.y - sh.y) / d;
            }
        }

    }


    /**
     * @param {number} timeStepMs 
     */
    move(timeStepMs = 10) {

        //compute new speed
        this.sx += this.ax * timeStepMs + (1 - 2 * Math.random()) * 0.02;
        this.sy += this.ay * timeStepMs + (1 - 2 * Math.random()) * 0.02;
        this.s = this.computeSpeed()
        if (this.s > this.sea.V_MAX) {
            this.s = this.sea.V_MAX;
            this.sx = this.sea.V_MAX * Math.cos(this.sa);
            this.sy = this.sea.V_MAX * Math.sin(this.sa);
        }

        //compute new position
        this.x += this.sx * timeStepMs;
        this.y += this.sy * timeStepMs;

        //limit
        if (this.x < 0) this.x = this.sea.w;
        if (this.y < 0) this.y = this.sea.h;
        if (this.x > this.sea.w) this.x = 0;
        if (this.y > this.sea.h) this.y = 0;
    }


    /**
     * Display a sardin body
     * 
     * @param {GeoViewer} cp 
     * @param {number} length 
     */
    display(cp, length) {
        const c2 = cp.ctx
        const c = Math.floor(255 * Math.abs(this.sa) / Math.PI);
        c2.strokeStyle = "rgb(255, 255, " + c + ")"
        const a = length / this.s;
        const dx = a * this.sx * 0.5;
        const dy = a * this.sy * 0.5;

        c2.beginPath();
        c2.moveTo(cp.geoToPixX(this.x - dx), cp.geoToPixY(this.y - dy));
        c2.lineTo(cp.geoToPixX(this.x + dx), cp.geoToPixY(this.y + dy));
        c2.closePath();
        c2.stroke();
    }

    /**
     * Display sardin vision field
     * 
     * @param {GeoViewer} cp 
     * @param {string} fillStyle 
     * @param {number} lineWidth 
     */
    displayVisionField(cp, fillStyle, lineWidth) {
        const c2 = cp.ctx
        c2.fillStyle = fillStyle;
        c2.lineWidth = lineWidth;

        c2.beginPath();
        c2.moveTo(cp.geoToPixX(this.x), cp.geoToPixY(this.y));
        c2.arc(cp.geoToPixX(this.x), cp.geoToPixY(this.y), this.sea.D_OBS / cp.getZf(),
            -this.sa - this.sea.A_OBS * 0.5,
            -this.sa + this.sea.A_OBS * 0.5
        );
        c2.closePath();
        c2.fill();
    }

    /**
     * Display sardin vision links
     * 
     * @param {GeoViewer} cp 
     * @param {string} strokeStyle 
     * @param {number} lineWidth 
     */
    displayVisionLinks(cp, strokeStyle, lineWidth) {
        const c2 = cp.ctx
        c2.strokeStyle = strokeStyle;
        c2.lineWidth = lineWidth;
        for (let sa2 of this.obs) {
            c2.beginPath();
            c2.moveTo(cp.geoToPixX(this.x), cp.geoToPixY(this.y));
            c2.lineTo(cp.geoToPixX(sa2.x), cp.geoToPixY(sa2.y));
            c2.closePath();
            c2.stroke();
        }
    }

    /**
     * Display sardin collision field
     * 
     * @param {GeoViewer} cp 
     * @param {string} strokeStyle 
     * @param {number} lineWidth 
     */
    displayCollisionField(cp, strokeStyle, lineWidth) {
        const c2 = cp.ctx
        c2.strokeStyle = strokeStyle;
        c2.lineWidth = lineWidth;
        c2.beginPath();
        c2.arc(cp.geoToPixX(this.x), cp.geoToPixY(this.y), this.sea.D_COL * 0.5 / cp.getZf(), 0, 2 * Math.PI);
        c2.closePath();
        c2.stroke();
    }

}
